### ISSUE TITLE
Feature request: socks5 frontend for client

### ISSUE THREAD

**Issue by rofl0r** [SoftEtherVPN/SoftEtherVPN#1369]
hello, thanks for providing this awesome product.

one thing that's been bothering me for quite some time with all VPN solutions is that they all require elevated privileges for creating tun/tap/bridge devices, and having to mess with the routing table, this makes usage of VPNs quite complex. in order to circumvent these issues, there exist dozens of docker images that encapsulate the VPN software into a container or VM, and provide a http or socks server to interface with the host. however this solution has its own drawbacks, such as high resource usage or requirement for elevated privileges for some container runtimes.

with a built-in socks5 or http proxy frontend for the client, no elevated privileges at all are required, and the vpn connection could be used without issues by regular users with proxy-aware applications like webbrowsers or IRC clients, and even with most non-proxy-aware applications using a preload hack like [proxychains-ng](https://github.com/rofl0r/proxychains-ng/), which would even allow it to chain the VPN with other proxy solutions. the client could even run in foreground mode without having different tools that communicate with background services.

example usage: `vpnsocksclient --listenip 127.0.0.1 --port 1080 --vpnconfig foo.conf`, client runs in foreground, and after usage the user can just hit CTRL-c and be done.

the only drawback is that it's TCP only, though this is sufficient for the vast majority of internet services.
the socks5 protocol is very simple and straightforward to implement, i've written one myself and the core is implemented in roughly 500 lines of C: https://github.com/rofl0r/microsocks/blob/master/sockssrv.c .

please let me know what you think about this proposal.

**Comment by davidebeatrici** [comment]
Hello, that's an excellent idea!

Right now the client service forces the association of a connection to a virtual network adapter, but removing the limitation is easy.

Implementing a SOCKS5 server should be easy as well, we already have a client: https://github.com/SoftEtherVPN/SoftEtherVPN/blob/f18837b49156bdc27e5b12485102dcdaf4315a91/src/Mayaqua/Proxy.c#L214-L525

**Comment by domosekai** [comment]
Socks server is probably not so hard to do. And Socks 5 actually has UDP support as well. 
However without the help of tap devices, do we have to do L2 to L3 bridging by ourselves? That would require lots of things like ARP, DHCP, ICMP, IPv6 PD etc. We are already doing a part of these at the server side but I'm not sure about the client.

**Comment by davidebeatrici** [comment]
I think the feature would only be used to route internet traffic towards the VPN server.

**Comment by rofl0r** [comment]
> And Socks 5 actually has UDP support as well.

that's true, however implementing it seems a lot more elaborate than TCP side of things. i for my part would be happy to have "just" TCP.

> However without the help of tap devices, do we have to do L2 to L3 bridging by ourselves? That would require lots of things like ARP, DHCP, ICMP, IPv6 PD etc.

from my understanding, you'd just get incoming requests to connect to a specific ipv4/v6 or hostname, establish the TCP connections, and then forward packets from both sides. i don't think we'd need to deal with any of ICMP, ARP, etc.

**Comment by domosekai** [comment]
Sorry I don't get it. Without the help of a virtual NIC, how can we assign an IP to the client? Without that, how can we make TCP connections?

**Comment by rofl0r** [comment]
> Without the help of a virtual NIC, how can we assign an IP to the client?

you don't. the VPN client uses the existing internet ip to connect to VPN server, then communicates with it as per VPN protocol. additionally, it creates a server socket listening on a specified port. the user then configures e.g. his browser to use that port as SOCKS proxy.
when data is received on the SOCKS port, it contains instructions to which IP/port tuple to connect to, so the VPN facing side of the code just tells the VPN server to establish a connection there, and then sends/forwards data received on that socket to the VPN server.

**Comment by domosekai** [comment]
What you are saying is encapsulating SOCKS within SE VPN procotol. But SE VPN is layer 2 which means the client is expected to send Ethernet packets to the server. It can't just send TCP payload which is layer 4. To fill the gap we need to maintain MAC / IP address even virtually.

**Comment by rofl0r** [comment]
when you look at a packet that is actually going out the WAN interface (using wireshark/tcpdump), it's probably an ordinary TCP packet, which contains inside of it a fake ethernet packet with ethernet header etc. the VPN client just has to store somewhere in its memory the IP it's been assigned to by the VPN server and doctor that into the fake ethernet packet sent to the destination (VPN server).
